src/saml.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the SamlApp class to encapsulate the business logic."""
+import logging
+import urllib
+from functools import cached_property
+from typing import List, Set, Optional
+
+from lxml import etree  # nosec
+
+from charm_state import CharmConfigInvalidError, CharmState
+
+from pydantic import AnyHttpUrl, BaseModel, Field
+
+logger = logging.getLogger(__name__)
+
+
+class SamlEndpoint(BaseModel):  # pylint: disable=too-few-public-methods
+    """Represent a SAML endpoint.
+
+    Attrs:
+        name: Endpoint name.
+        url: Endpoint URL.
+        binding: Endpoint binding.
+        response_url: URL to address the response to.
+    """
+
+    name: str = Field(..., min_length=1)
+    url: AnyHttpUrl
+    binding: str = Field(..., min_length=1)
+    response_url: Optional[AnyHttpUrl]
+
+
+class SamlIntegrator:  # pylint: disable=too-few-public-methods
+    """A class representing the SAML Integrator application.
+
+    Attrs:
+        entity_id: Entity ID for SAML.
+        metadata_url: URL for the SAML metadata.
+        endpoints: SAML endpoints.
+        certificates: Public certificates.
+    """
+
+    def __init__(self, charm_state: CharmState):
+        """Initialize a new instance of the SamlApp class.
+
+        Args:
+            charm_state: The state of the charm that the Saml instance belongs to.
+        """
+        self._charm_state = charm_state
+
+    @property
+    def entity_id(self) -> str:
+        """Return entity_id config.
+
+        Returns:
+            str: entity_id config.
+        """
+        return self._charm_state.entity_id
```

Comment:
This kind of looks the same as the state?

Reply:
Yes, this wraps the state properties. Do you have any change to suggest?

Reply:
What is the reason for needing the wrapper?

Reply:
E.g., the saml functions could just take the state as an argument

Reply:
Can you give me an example of that?

Reply:
something like this:

```python
def function_requiring_state(state: State): -> ...:
    pass
```

Reply:
This entity is a candidate to be moved to the library anyway as both the provider and the requirer will needed. I suggest we address this in the lib PR. thanks for the input